### DISCUSSION TITLE
Mise fails to install nested Go tools

### DISCUSSION THREAD

**Discussion by zx8** [jdx/mise#3822] [Troubleshooting and bug reports]
**Describe the bug**

Mise seems to sometimes(?) fail to install Go tools that are in subdirectory of the repository.

**To Reproduce**

```
$ mise install go:github.com/noperator/sol/cmd/sol@latest
go: invalid github.com import path "github.com/noperator"
go: invalid github.com import path "github.com"
mise WARN  No versions found for go:github.com/noperator/sol/cmd/sol
```

**Expected behavior**

It works, like `go install`.

**`mise doctor` output**

```text
version: 2024.12.19 macos-arm64 (2024-12-23)
activated: yes
shims_on_path: no

build_info:
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS
  Built: Mon, 23 Dec 2024 19:02:19 +0000
  Rust Version: rustc 1.83.0 (90b35a623 2024-11-26) (Homebrew)
  Profile: release

shell:
  /opt/homebrew/bin/fish
  fish, version 3.7.1-2640-gf3dd4ee02

dirs:
  cache: ~/Library/Caches/mise
  config: ~/.config/mise
  data: ~/.local/share/mise
  shims: ~/.local/share/mise/shims
  state: ~/.local/state/mise

config_files:
  ~/.config/mise/config.toml

ignored_config_files: (none)

backends:
  aqua
  asdf
  cargo
  core
  dotnet
  gem
  go
  npm
  pipx
  spm
  ubi
  vfox

plugins:
  terraform           https://github.com/asdf-community/asdf-hashicorp.git#f1602b6
  usage               https://github.com/jdx/mise-usage.git#fe3888a

toolset:
  go:golang.org/x/tools/gopls@0.17.1
  go:mvdan.cc/gofumpt@0.7.0

env_vars:
  MISE_SHELL=fish

settings:
  experimental         true  ~/.config/mise/config.toml
  pipx.uvx             true  ~/.config/mise/config.toml
  python.uv_venv_auto  true  ~/.config/mise/config.toml
  status.show_env      true  ~/.config/mise/config.toml
  status.show_tools    false ~/.config/mise/config.toml

No problems found
```

**Comment by eitamal** [comment]
That's an interesting find! But, I think it still completed successfully, no? The main issue seems to do with the lack of versions for `sol`. But I think this behaviour is actually expected due to how Go modules work with repositories that don't have version tags.

While `go install github.com/noperator/sol/cmd/sol@latest` works because Go can fall back to [pseudo-versioning](https://go.dev/doc/modules/version-numbers#pseudo-version-number), the `go list -m -versions` command (which `mise` uses internally to discover available versions) doesn't return any results for repositories without explicit version tags.

You can verify this by running:
```sh
$ go list -m -versions -json github.com/noperator/sol/cmd/sol
{
        "Path": "github.com/noperator/sol/cmd/sol"
}
```
versus a repo with version tags:
```sh
$ go list -m -versions -json github.com/DarthSim/hivemind
{
        "Path": "github.com/DarthSim/hivemind",
        "Versions": [
                "v1.0.1",
                "v1.0.2",
                "v1.0.3",
                "v1.0.4",
                "v1.0.5",
                "v1.0.6",
                "v1.1.0"
        ]
}
```

So yeah, I think showing a warning is the correct thing to do here. The `go: invalid github...` is probably not great, though.

But the following works for me:
```sh
$ mise x go:github.com/noperator/sol/cmd/sol@latest -- sol --help
Usage of sol:
  -a    arguments
  -all
        all
  -b    binary commands: &&, ||, |, |&
  -c    command substitution: $(), ``
  -e    inspect env to resolve command types
  -f string
        file
  -j    jq
  -jqarr
        arrays
  -jqobj
        objects
  -jqop string
        operators (comma-separated)
  -l    clauses: case, for, if, while
  -o    one line
  -p    process substitution: <(), >()
  -r    redirect: >, >>, <, <>, <&, >&, >|, <<, <<-, <<<, &>, &>>
  -s    shell strings: xargs, parallel
  -v    verbose
  -w int
        max line width before breaking
```

_NOTE: I'm **not a maintainer**; just giving my two cents_ 😅

**Comment by jbduncan** [comment]
I'm seeing something similar when updating `github.com/tailscale/depaware`, which also doesn't have proper versions, in my project: https://github.com/jbduncan/go-containers:

```
$ mise up
go: invalid github.com import path "github.com/tailscale"
go: invalid github.com import path "github.com"
mise WARN  No versions found for go:github.com/tailscale/depaware
go: unrecognized import path "go.uber.org": parse https://go.uber.org/?go-get=1: no go-import meta tags ()
mise WARN  Error getting latest version for go:go.uber.org/nilaway/cmd/nilaway: timed out: timed out waiting on channel
mise WARN  Error getting latest version for go:golang.org/x/tools/cmd/eg: timed out: timed out waiting on channel
```

I think the two bottom warnings are happening for different reasons, so I'll report them separately.

**Comment by JamBalaya56562** [comment]
**This is fixed.** 👍 (@eitamal diagnosed it correctly)

For a repo with no version tags, `go list -m -versions` returns nothing, so mise reported `No versions found` (with the noisy `invalid github.com import path` lines Go prints while the path prefixes are probed). `go install` works because it falls back to a [pseudo-version](https://go.dev/doc/modules/version-numbers#pseudo-version-number) — mise now does the same.

**Fixed by #9097** ("fix(go): honor install_before for module versions"), which makes the go backend fall back to the **Go module proxy** for version metadata and returns the latest pseudo-version when there are no tags. First released in **v2026.4.14** (2026-04-15). I bisected it to confirm: v2026.4.12 still fails, v2026.4.14 works.

Verified on the current release (`2026.7.12`) — both cases from this thread now resolve to a pseudo-version instead of erroring:
```
$ mise ls-remote go:github.com/noperator/sol/cmd/sol
0.0.0-20240903161619-7762c5115dd8
$ mise ls-remote go:github.com/tailscale/depaware        # @jbduncan's case
0.0.0-20260720165112-f20f66241ec6
```
`mise install go:github.com/noperator/sol/cmd/sol@latest` installs cleanly now. (@jbduncan — the separate `unrecognized import path "go.uber.org"` line was a transitive-dependency lookup, not this bug.) Upgrading to v2026.4.14+ resolves it — could you confirm so this can be closed?

*This comment was generated by an AI coding assistant.*

**Reply by jbduncan** [reply]
@JamBalaya56562 Yep, I can confirm that the issue with the `unrecognized import path "go.uber.org"` line is fixed. Thank you very much for letting me know. :)
